magento/magento2-page-builder
# Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option

### Description (*)
Adding a bundle product to the cart from the product carousel/widget doesn’t let the product be added properly with the bundle option. The bundle product options are not added within the parent product. It shows a warning message in minicart like ‘Please specify product option(s)’. It doesn’t fail in all the cases, it only fails in a certain case where a bundle product has been created with a required bundle option & the bundle option contains only one product with the ‘is_default’ checkbox selected.

Screenshot 1:
![image](https://github.com/magento/magento2-page-builder/assets/106948523/09cd00ac-1166-4170-b8b1-1b497cf85870)
Here you can see in the screenshot a bundle product has been created with a bundle option containing only one product in it & the option is required and the product is default select.

Screenshot 2:
![admin-carousel](https://github.com/magento/magento2-page-builder/assets/106948523/7c48c08d-17d7-4fa7-9a6c-bd2ab9fcc07d)
After creating a bundle product, a product carousel has been added to the home page content by using a page-builder that includes the particular bundle product.

Screenshot 3:
![home-page](https://github.com/magento/magento2-page-builder/assets/106948523/0d10db8c-080c-4b7d-80e0-9d29539a1fc7)
As we have created a product carousel, a bundle product is visible on the home page in the product slider.

Screenshot 4:
![mini-cart](https://github.com/magento/magento2-page-builder/assets/106948523/902d9c09-cd17-49a9-9eff-983fe353ef76)
When we click on the add-to-cart button, the product seems to be added to the cart & it doesn’t throw any error. And then if you check in the mini-cart, you will see a warning message like “Please specify product option(s).”

### Related Pull Requests

1. https://github.com/magento/magento2/pull/38601

### Manual testing scenarios (*)
1. Create a bundle product that contains one or more bundle options & each bundle option is required and must contain only one product that is default selected. (see screenshot 1)
2. Create a product carousel using page-builder content that includes the newly created bundle product in the widget. (see screenshot 2). Here I've added the product carousel in home page content
3. From the front end visit the page & try to add the bundle product from the product slider.(see screenshot 3)
4. After clicking Add to Cart observe a warning message in minicart. (see screenshot 4).

### Expected result (*)
The bundle product should be added to the cart with bundle options.

### Actual result (*)
The bundle product isn’t being added to the cart with bundle options.

### Additional Information (*)
File: `Magento/PageBuilder/view/frontend/templates/catalog/product/widget/content/carousel.phtml`

In the given `carousel.phtml` file it doesn’t get the options data where the add-to-cart form is rendered for the particular product (see screenshot 5). As the options are not rendered, adding that product to the cart will only add the parent bundle product to the cart without any bundle options.

Screenshot 5:
![code-issue](https://github.com/magento/magento2-page-builder/assets/106948523/dd302a4c-5cdc-4abc-a7ea-79dc7ff986d2)

### Solution (*)
We have to update the given `carousel.phtml`  to get the bundle options in the add-to-cart form. We can get the bundle product options by using the `Magento\Catalog\ViewModel\Product\OptionsData` ViewModel. (see  screenshot 6)

A separate function has been created in the block file of the `carousel.phtml` (`Magento\CatalogWidget\Block\Product\ProductsList`) to get the ViewModel that returns the object of `Magento\Catalog\ViewModel\Product\OptionsData`. Refer the mentioned related PR.

Screenshot 7:
![code-resolve](https://github.com/magento/magento2-page-builder/assets/106948523/4cf783f9-b773-43c5-9771-6a433186a2bc)

The same issue happens with the product grid as well in the product slider added by page builder. That can be solved by this PR:

- https://github.com/magento/magento2/pull/38601

### Checklist


### Resolved issues:
1. [x] resolves magento/magento2-page-builder#870: Fixes: Bundle product with only one product in required bundle option not being added to cart with bundle option
